substrate/frame/revive/src/evm/runtime.rs

```diff
@@ -0,0 +1,638 @@
+		BalanceOf<Self::Config>: Into<U256> + TryFrom<U256>,
+		MomentOf<Self::Config>: Into<U256>,
+		AccountIdOf<Self::Config>: Into<AccountId32>,
+		<Self::Config as frame_system::Config>::RuntimeCall: Dispatchable<Info = DispatchInfo>,
+		OnChargeTransactionBalanceOf<Self::Config>: Into<BalanceOf<Self::Config>>,
+		Call: From<crate::Call<Self::Config>> + TryInto<crate::Call<Self::Config>>,
+	{
+		let tx = rlp::decode::<TransactionLegacySigned>(&payload).map_err(|err| {
+			log::debug!(target: LOG_TARGET, "Failed to decode transaction: {err:?}");
+			InvalidTransaction::Call
+		})?;
+
+		let signer = tx.recover_eth_address().map_err(|err| {
+			log::debug!(target: LOG_TARGET, "Failed to recover signer: {err:?}");
+			InvalidTransaction::BadProof
+		})?;
+
+		let signer =
+			<Self::Config as crate::Config>::AddressMapper::to_account_id_contract(&signer);
+		let TransactionLegacyUnsigned { nonce, chain_id, to, value, input, gas, gas_price, .. } =
+			tx.transaction_legacy_unsigned;
+
+		if chain_id.unwrap_or_default() != <Self::Config as crate::Config>::ChainId::get().into() {
+			log::debug!(target: LOG_TARGET, "Invalid chain_id {chain_id:?}");
+			return Err(InvalidTransaction::Call);
+		}
+
+		let call = if let Some(dest) = to {
+			if !matches!(transact_kind, EthTransactKind::Call) {
+				log::debug!(target: LOG_TARGET, "Invalid transact_kind, expected Call");
+				return Err(InvalidTransaction::Call);
+			}
+
+			crate::Call::call::<Self::Config> {
+				dest,
+				value: value.try_into().map_err(|_| InvalidTransaction::Call)?,
+				gas_limit,
+				storage_deposit_limit,
+				data: input.0,
+			}
+		} else {
+			let EthTransactKind::InstantiateWithCode { code_len, data_len } = transact_kind else {
+				log::debug!(target: LOG_TARGET, "Invalid transact_kind, expected InstantiateWithCode");
+				return Err(InvalidTransaction::Call);
+			};
+
+			let blob = match polkavm::ProgramBlob::blob_length(&input.0) {
+				Some(blob_len) => blob_len
+					.try_into()
+					.ok()
+					.and_then(|blob_len| (input.0.split_at_checked(blob_len))),
+				_ => None,
+			};
+
+			let Some((code, data)) = blob else {
+				log::debug!(target: LOG_TARGET, "Failed to extract polkavm code & data");
+				return Err(InvalidTransaction::Call);
+			};
+
+			if code.len() as u32 != code_len || data.len() as u32 != data_len {
+				log::debug!(target: LOG_TARGET, "Invalid code or data length");
+				return Err(InvalidTransaction::Call);
+			}
+
+			crate::Call::instantiate_with_code::<Self::Config> {
+				value: value.try_into().map_err(|_| InvalidTransaction::Call)?,
+				gas_limit,
+				storage_deposit_limit,
+				code: code.to_vec(),
+				data: data.to_vec(),
+				salt: None,
+			}
+		};
+
+		let nonce = nonce.try_into().map_err(|_| InvalidTransaction::Call)?;
+		let eth_fee =
+			gas_price.saturating_mul(gas).try_into().map_err(|_| InvalidTransaction::Call)?;
+		let info = call.get_dispatch_info();
+		let function: Call = call.into();
+
+		let tip = Default::default();
```

Comment:
When `eth_fee` is larger than `actual_fee` we probably should supply the difference as `tip`. This would enable people to prioritise their transaction by supplying a higher gas fee. Just as they would on eth. Not sure how they determine the `gas_price` in the first place, though.

Reply:
we still want to ensure that the injected weight match the fees the user specified in the transaction.
Since we hard code the gas_price to 1,  we can do our check using this value, and if the user provided a gas_price greater than 1,  use the difference as tip.

for post eip1559 tx types there is a maxPriorityFeePerGas that define the tip value

will amend to add the tip calculation.